custom/abt/reports/late_pmt.py

```diff
@@ -204,17 +204,15 @@ def _to_report_format(date, user, group):
         rows = []
         sub_status = self.report_config['submission_status']
         if users:
-            group_a = self.get_users_in_group_a
-            group_b = self.get_users_in_group_b
-
             for date in dates:
                 for user in users:
-                    key = (date.date(), user['user_id'])
-                    if key not in group_a and sub_status != 'group_b':
-                        group = _('No PMT data Submitted')
-                    elif key not in group_b and key in group_a and sub_status != 'group_a':
-                        group = _('Incorrect PMT data Submitted')
+                    sms_received = (date.date(), user['user_id']) in self.smss_received
+                    sms_event_exists = (date.date(), user['user_id']) in self.sms_events
+                    if not sms_received and sub_status != 'group_b':
+                        error_msg = _('No PMT data Submitted')
+                    elif not sms_event_exists and sms_received and sub_status != 'group_a':
+                        error_msg = _('Incorrect PMT data Submitted')
                     else:
                         continue
-                    rows.append(_to_report_format(date, user, group))
+                    rows.append(_to_report_format(date, user, error_msg))
```

Comment:
I think this could be replaced with the following conditional, though not confident enough to make the change right now:

```python
    if not sms_received:
        if sub_status != 'group_b':
            rows.append(_to_report_format(date, user, _('No PMT data Submitted')))
    elif not sms_event_exists:
        if sub_status != 'group_a':
            rows.append(_to_report_format(date, user, _('Incorrect PMT data Submitted')))
```

Reply:
```python
    elif not sms_event_exists:
        if sub_status != 'group_a':
```
If understand this correctly, the code is trying to check "if we didn't send an SMS but we did receive an SMS ... and if the user is not filtering by status SMS_RECEIVED_FROM_MW" (then the SMS probably doesn't contain data that the user is looking for).

I could be wrong about that. But if I'm right, I think you probably need both conditions:
```python
    elif sms_received and not sms_event_exists:
```

Reply:
the `sms_received` is implicit.  If `sms_received` were falsy, it wouldn't reach this `elif` conditional.

I think this would be further clarified by adding an enriching phase before this where we categorize the data as "fine", "missing", or "incorrect", then here decide whether to filter out or add an error message as appropriate.

Reply:
> If `sms_received` were falsy, it wouldn't reach this elif conditional.

Oh, of course.

> where we categorize the data as "fine", "missing", or "incorrect"

That sounds like a great idea.